feat: Support driver pod labels and annotations configuration at API server startup

thc1006 opened the pull request:
## Summary

This PR implements admin-level configuration for driver pod labels and annotations, addressing issue #12015. Configuration is loaded once at API server startup and cached for performance.

### Key Changes

- **Startup-time configuration loading**: Driver pod config (labels and annotations) is initialized once when the API server starts via `common.InitDriverPodConfig()` in backend/src/apiserver/main.go
- **Thread-safe caching**: Uses `sync.RWMutex` for safe concurrent access to cached configuration
- **Reserved prefix filtering**: Automatically filters out labels and annotations with the `pipelines.kubeflow.org/` prefix to prevent overriding system-managed metadata
- **JSON configuration format**: Supports JSON format for configuration (as per reviewer feedback)
- **Comprehensive testing**: Unit tests covering all configuration scenarios including JSON string input from ConfigMap

### All Reviewer Requirements Addressed (7/7)

✅ **Removed .gitignore changes** - Only relevant backend and manifest changes included  
✅ **Using Viper configuration** - Integrated with existing Viper config system  
✅ **JSON format only** - Removed YAML support, JSON format only  
✅ **Kubernetes-compatible format** - Standard Kubernetes label/annotation format  
✅ **Startup-time loading with caching** - Config loaded once at startup, cached for performance  
✅ **Documentation in backend/README.md** - Concise documentation added  
✅ **Removed unimplemented options** - Removed environment variable and serviceaccount options

### Testing

Local tests passed:
- ✅ All driver config unit tests (backend/src/apiserver/common/driver_config_test.go)
- ✅ All argocompiler unit tests  
- ✅ go mod tidy
- ✅ go vet
- ✅ gofmt

### Configuration Example

```json
{
  "DriverPodLabels": {
    "custom-label": "value",
    "team": "ml-platform"
  },
  "DriverPodAnnotations": {
    "sidecar.istio.io/inject": "true",
    "custom-annotation": "value"
  }
}
```

Resolves: #12015

---

**Note**: This PR supersedes #12306 which was closed due to commit history issues. All code changes and reviewer feedback remain fully addressed in this clean PR.

thc1006 committed:
feat: support driver pod labels and annotations configuration at API server startup

Enable admin-level configuration for driver pod labels and annotations,
addressing issue #12015. Configuration is loaded once at API server startup
via Viper and cached with thread-safe access for performance.

Key changes:
- Wire DRIVER_POD_LABELS and DRIVER_POD_ANNOTATIONS from the pipeline-install-config
  ConfigMap into the API server Deployment as environment variables, so the ConfigMap
  path documented in the README actually reaches Viper. This is done for the default
  install and for the PostgreSQL install, whose apiserver patch replaces the container
  environment list and therefore needs its own entries.
- Add InitDriverPodConfig() to load, validate, and cache config at startup. Every value
  must be a JSON string. A malformed document, a boolean, a number and a JSON null are
  all rejected with a descriptive error instead of being silently dropped or coerced to
  an empty string by viper.GetStringMapString.
- Reject label keys, label values, and annotation keys that Kubernetes would not accept,
  so a bad entry fails at startup instead of when the driver pod is created. Annotation
  values stay free form so inline JSON such as the Istio proxy config keeps working.
- Cache the configuration behind a single struct pointer swapped atomically, and copy
  both maps under one read lock so a reader can never mix labels from one version of the
  config with annotations from another.
- Pass the configuration into the compiler through argocompiler.Options, following the
  existing DefaultRunAs path, so the compiler does not depend on API server state and
  callers such as the standalone compiler simply leave it nil.
- Filter the reserved pipelines.kubeflow.org/ prefix from both labels and annotations to
  protect system-managed metadata, and apply the config to both the DAG and container
  driver pod templates while preserving any existing system-set keys.
- Tests cover the environment variable contract the Deployment relies on, the empty
  default shipped in every manifest, every rejected value type including null, metadata
  format validation, reserved-prefix filtering, copy isolation, and the full compile path
  asserting the config lands on the generated driver templates.

Signed-off-by: thc1006 <a0989246280@gmail.com>
Signed-off-by: thc1006 <84045975+thc1006@users.noreply.github.com>

hbelmiro commented on backend/src/apiserver/common/driver_config.go:
```diff
@@ -0,0 +1,135 @@
+	}
+
+	// Load labels from Viper configuration
+	rawLabels := GetMapConfig(DriverPodLabels)
```
`InitDriverPodConfig()` claims to "catch configuration errors early" (line 649 in `main.go`), but `GetMapConfig` delegates to `viper.GetStringMapString`, which silently returns an empty map on the **ConfigMap string path** when it cannot parse the JSON string into `map[string]string`. Note: config.json native map values are **not** affected — Viper coerces non-string values (booleans, numbers) to strings when the source is a `map[string]interface{}`. The silent failure is specific to the ConfigMap path where values arrive as JSON strings, which is the documented deployment method in both the PR's README and manifest. Verified against Viper:

| Source / value                                    | Result                                           |
|---------------------------------------------------|--------------------------------------------------|
| ConfigMap: `'{"sidecar.istio.io/inject":"true"}'` | `map[sidecar.istio.io/inject:true]`              |
| ConfigMap: `'{"sidecar.istio.io/inject":true}'`   | `map[]` — boolean instead of string              |
| ConfigMap: `'{"replicas":3}'`                     | `map[]` — number instead of string               |
| ConfigMap: `'{bad json}'`                         | `map[]` — syntax error                           |
| config.json: `{"sidecar.istio.io/inject": true}`  | `map[sidecar.istio.io/inject:true]` — coerced OK |

An operator who writes `true` (JSON boolean) instead of `"true"` (JSON string) for the Istio sidecar injection value in the ConfigMap — a natural mistake — sees `"Driver pod configuration initialized: 0 labels, 0 annotations"` at INFO level and believes nothing is configured, when the configuration is silently broken.

**Why it matters:** The entire rationale for startup-time loading (stated in code comments and the PR description) is to surface configuration problems early. Silent swallowing defeats that purpose and makes misconfiguration hard to diagnose in production.

**Suggested fix:** Change `InitDriverPodConfig` to return `error` and fail startup on bad config. `initConfig()` in `main.go` already returns `error`, so this propagates naturally. After calling `GetMapConfig`, probe the raw string with `json.Unmarshal` when the parsed map is empty. `GetMapConfig` returns an empty non-nil map (not `nil`) for unparseable input, so a nil check is insufficient. A simple `len() == 0` heuristic would false-positive on valid `"{}"`, so an explicit parse check is more precise:

```go
func InitDriverPodConfig() error {
	driverConfigMutex.Lock()
	defer driverConfigMutex.Unlock()

	// ... existing re-init log ...

	rawLabels := GetMapConfig(DriverPodLabels)
	if err := validateMapConfig(DriverPodLabels, rawLabels); err != nil {
		return err
	}
	cachedDriverPodLabels = filterReservedEntries(rawLabels)

	rawAnnotations := GetMapConfig(DriverPodAnnotations)
	if err := validateMapConfig(DriverPodAnnotations, rawAnnotations); err != nil {
		return err
	}
	cachedDriverPodAnnotations = filterReservedEntries(rawAnnotations)

	driverConfigInitialized = true
	// ... existing logging ...
	return nil
}

func validateMapConfig(name string, parsed map[string]string) error {
	if len(parsed) > 0 || !viper.IsSet(name) {
		return nil
	}
	raw := viper.GetString(name)
	if raw == "" {
		return nil
	}
	var probe map[string]interface{}
	if err := json.Unmarshal([]byte(raw), &probe); err != nil {
		return fmt.Errorf("%s could not be parsed as JSON — check syntax: %w (raw value: %q)", name, err, raw)
	}
	if len(probe) > 0 {
		return fmt.Errorf("%s parsed as JSON but all values were dropped — ensure every value is a string, not a boolean or number (raw value: %q)", name, raw)
	}
	return nil
}
```

Update the call site in `main.go`:

```go
if err := common.InitDriverPodConfig(); err != nil {
	return fmt.Errorf("driver pod config: %w", err)
}
```

thc1006 replied:
Thank you so much for catching this, and for the clear table showing how Viper treats each value. You are right. I changed `InitDriverPodConfig` to return an error, and `initConfig` in `main.go` now passes it up, so the server refuses to start on a broken configuration instead of coming up silently empty. I added a helper `validateMapConfig` that inspects the raw value again with `json.Unmarshal` whenever `GetStringMapString` returns an empty map, which is exactly the approach you proposed. I also reproduced the behaviour locally to be sure: a boolean value, a numeric value, and invalid JSON all collapse to an empty map, while a valid empty object is treated as a genuine empty configuration rather than an error. I really appreciate the care you put into writing this up.

hbelmiro commented on backend/src/apiserver/common/driver_config_test.go:
No test covers the ConfigMap failure paths identified in the previous comment. The test suite validates empty config, valid maps, valid JSON strings, reserved prefix filtering, and copy isolation — but never asserts that `InitDriverPodConfig` returns an error when Viper receives a non-string JSON value (e.g., boolean `true` instead of string `"true"`) or syntactically invalid JSON via the ConfigMap string path.

**Suggested fix:** Add test cases for both failure modes, asserting that startup-time validation surfaces the error:

```go
func TestInitDriverPodConfig_NonStringJSONValues(t *testing.T) {
	viper.Reset()
	resetDriverConfig()

	// Boolean instead of string — the most realistic ConfigMap typo
	viper.Set(DriverPodLabels, `{"sidecar.istio.io/inject":true}`)

	err := InitDriverPodConfig()

	require.Error(t, err, "non-string JSON values via ConfigMap should fail at startup")
	assert.Contains(t, err.Error(), "DriverPodLabels")
}

func TestInitDriverPodConfig_MalformedJSON(t *testing.T) {
	viper.Reset()
	resetDriverConfig()

	viper.Set(DriverPodLabels, "{bad json}")

	err := InitDriverPodConfig()

	require.Error(t, err, "malformed JSON should fail at startup")
	assert.Contains(t, err.Error(), "DriverPodLabels")
}
```

These verify that operators see a clear startup failure instead of a silent no-op.

thc1006 replied:
Good call, thank you. I added two tests for the failure paths you described. `TestInitDriverPodConfig_NonStringJSONValues` covers a boolean where a string is expected, and `TestInitDriverPodConfig_MalformedJSON` covers invalid syntax. Both assert that the returned error names the offending key so an operator can see what to correct. I also added `TestInitDriverPodConfig_EmptyJSONObjectIsValid` to keep a guard proving that a valid empty object does not fail startup.

hbelmiro commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -275,6 +289,9 @@ func (c *workflowCompiler) addContainerDriverTemplate() string {
 	}
 	applySecurityContextToTemplate(template)
+
+	c.driverPodConfig.applyToTemplate(template)
```
Tests cover `applyToTemplate` and `getDriverPodConfig` individually, but no test verifies the full compilation path: `Compile()` → `addContainerDriverTemplate()` / `addDAGDriverTemplate()` producing a workflow with the expected labels and annotations on driver templates. The existing `TestContainerDriverTemplate_IncludesKFPPodNameEnv` test (container_test.go:73) demonstrates exactly how to test this pattern — create a `workflowCompiler` with a `driverPodConfig`, call `addContainerDriverTemplate()`, and assert the resulting template's metadata.

**Why it matters:** Unit tests for `applyToTemplate` prove the method works in isolation, but don't guarantee the integration point in `addContainerDriverTemplate` and `addDAGDriverTemplate` actually calls it at the right time (after template creation, before template storage). A refactor that reorders those calls could silently break the feature.

**Suggested fix:** Add a test similar to `TestContainerDriverTemplate_IncludesKFPPodNameEnv` that sets `driverPodConfig` on the compiler and verifies labels appear in the generated template:

```go
func TestContainerDriverTemplate_IncludesDriverPodConfig(t *testing.T) {
	proxy.InitializeConfigWithEmptyForTests()
	c := &workflowCompiler{
		templates: make(map[string]*wfapi.Template),
		wf:        &wfapi.Workflow{Spec: wfapi.WorkflowSpec{Templates: []wfapi.Template{}}},
		spec:      &pipelinespec.PipelineSpec{PipelineInfo: &pipelinespec.PipelineInfo{Name: "test"}},
		job:       &pipelinespec.PipelineJob{},
		driverPodConfig: &driverPodConfig{
			Labels:      map[string]string{"sidecar.istio.io/inject": "true"},
			Annotations: map[string]string{"proxy.istio.io/config": "hold"},
		},
	}
	name := c.addContainerDriverTemplate()
	tmpl := c.templates[name]
	require.NotNil(t, tmpl)
	assert.Equal(t, "true", tmpl.Metadata.Labels["sidecar.istio.io/inject"])
	assert.Equal(t, "hold", tmpl.Metadata.Annotations["proxy.istio.io/config"])
}
```

thc1006 replied:
Thank you for pointing me to `TestContainerDriverTemplate_IncludesKFPPodNameEnv` as a model, that made it straightforward. I added `TestContainerDriverTemplate_IncludesDriverPodConfig`, and a matching `TestDAGDriverTemplate_IncludesDriverPodConfig` for the DAG driver, since the same reasoning applies to both templates. Each test builds a `workflowCompiler` with a configuration set, calls the real `addContainerDriverTemplate` or `addDAGDriverTemplate`, and asserts that the labels and annotations end up on the generated template metadata. A future change to the order of those calls would now be caught.

hbelmiro commented on backend/src/apiserver/common/driver_config.go:
```diff
@@ -0,0 +1,135 @@
+	ReservedLabelPrefix = "pipelines.kubeflow.org/"
+)
+
+var (
```
The cached driver pod configuration is spread across three independent package-level variables (`cachedDriverPodLabels`, `cachedDriverPodAnnotations`, `driverConfigInitialized`) that must stay in sync but are mutated one at a time. This makes it easy to leave the state half-updated — e.g., labels written but annotations not yet, or `driverConfigInitialized` set before both maps are ready. The `InitDriverPodConfig` function currently doesn't return errors, but if error handling is added (see Comment 1), early returns would leave these variables inconsistent.

**Suggested fix:** Group the cached state into a struct with a constructor that either returns a fully valid object or an error. The package-level state becomes a single pointer swapped atomically:

```go
type driverPodConfigCache struct {
	labels      map[string]string
	annotations map[string]string
}

var (
	cachedDriverPodConfig *driverPodConfigCache
	driverConfigMutex     sync.RWMutex
)

func InitDriverPodConfig() error {
	// Build and validate before taking the lock for the swap.
	cfg, err := newDriverPodConfigCache()
	if err != nil {
		return err
	}

	driverConfigMutex.Lock()
	defer driverConfigMutex.Unlock()
	cachedDriverPodConfig = cfg
	return nil
}

func newDriverPodConfigCache() (*driverPodConfigCache, error) {
	rawLabels := GetMapConfig(DriverPodLabels)
	if err := validateMapConfig(DriverPodLabels, rawLabels); err != nil {
		return nil, err
	}
	rawAnnotations := GetMapConfig(DriverPodAnnotations)
	if err := validateMapConfig(DriverPodAnnotations, rawAnnotations); err != nil {
		return nil, err
	}
	return &driverPodConfigCache{
		labels:      filterReservedEntries(rawLabels),
		annotations: filterReservedEntries(rawAnnotations),
	}, nil
}

func GetDriverPodLabels() map[string]string {
	driverConfigMutex.RLock()
	defer driverConfigMutex.RUnlock()
	if cachedDriverPodConfig == nil {
		return nil
	}
	return copyMap(cachedDriverPodConfig.labels)
}
```

This eliminates the `driverConfigInitialized` flag (nil check on the struct replaces it), guarantees all-or-nothing state transitions, and keeps validation/construction separate from the cache swap.

thc1006 replied:
Agreed, and thank you for the suggestion. This is much cleaner. I grouped the cached state into a single `DriverPodConfig` value held behind one pointer guarded by the mutex. A new helper `newDriverPodConfig` reads and validates everything first, and `InitDriverPodConfig` then swaps the pointer in one step under the lock, so a reader never sees a configuration that is only partly populated. The `driverConfigInitialized` flag is gone, since a nil pointer already tells us the configuration has not been loaded yet.

hbelmiro commented on backend/src/v2/compiler/argocompiler/argo.go:
```diff
@@ -318,6 +319,41 @@ type workflowCompiler struct {
+}
+
+// driverPodConfig holds labels and annotations to be applied to driver pods
+type driverPodConfig struct {
```
Once you create a struct as suggested in the previous comment, this one becomes duplicated.

thc1006 replied:
Thank you, you are right that it would otherwise be duplicated. I removed the `driverPodConfig` struct from the argo compiler. The compiler now uses the exported `common.DriverPodConfig` directly, and the old `applyToTemplate` method became a small standalone function `applyDriverPodConfig`, so only one type describes this configuration now.

thc1006 commented:
Thank you very much for the thorough and thoughtful review. I have addressed all five of your comments and rebased the branch onto the latest master, which also clears the earlier merge conflict. Locally the `common` and `argocompiler` packages pass build, vet and their tests including the new cases, and I confirmed that the `apiserver` package still compiles with CGO enabled. Please let me know if you would like any further changes, and thank you again for your time.

hbelmiro commented:
/ok-to-test

hbelmiro commented on frontend/server/package-lock.json:
These changes seem unrelated.

thc1006 replied:
Thank you, you are right. This was an unrelated npm lockfile change that slipped into the commit by accident. I have removed it and rebased onto the latest master, so this file is no longer part of the PR.

hbelmiro commented on test/frontend-integration-test/package-lock.json:
These changes seem unrelated.

thc1006 replied:
Agreed, thank you. Same here, this lockfile change was unrelated and got committed by mistake. It is now reverted, so the diff only contains the driver pod configuration changes again.

hbelmiro commented:
/ok-to-test

thc1006 commented:
The only failing job is the MLflow scenario with cacheEnabled=false and artifactStorage=file. The other three MLflow matrix variants passed on this exact commit, so this looks like a flaky issue in the MLflow operator deploy and port forward rather than anything in this change. This change only adds labels and annotations to driver pods when an admin configures them, and these tests leave that configuration empty, so the generated workflows are unchanged. Kindly requesting a retest when convenient, and thank you.

/retest

hbelmiro reviewed (approved):
/lgtm
/approve

cc @juliusvonkohout for approval

thc1006 commented:
Thank you so much @hbelmiro for the review, the lgtm, the approve, and for cc'ing @juliusvonkohout. I really appreciate you helping move this along.

The only remaining step is an approval from the manifests/kustomize OWNERS, since the PR adds two ConfigMap keys there. @juliusvonkohout or @chensun, whenever you have a moment, I would be very grateful for a look. CI is green and the branch is currently free of conflicts, so I am hoping to land it before master moves again. Thank you all very much.

hbelmiro commented:
@chensun the GitHub approval no longer adds the `approved` label.
You need to comment `/approve` now.

thc1006 commented:
Thank you so much @chensun for reviewing and approving, and @hbelmiro for helping route this. I understand the Prow approved label now needs a /approve comment rather than the review button, so whenever you have a moment @chensun I would be grateful for that. CI is green and the branch is still free of conflicts. Thank you both very much.

thc1006 commented:
Before this merged I wanted to be sure the ConfigMap path actually worked end to end, and I am glad I checked, because it did not. I would rather hold the PR up now than let operators discover it afterwards.

@hbelmiro, it was your earlier point about the ConfigMap string path being the documented way to deploy this that sent me back to look at how those keys reach the API server. They never reached it. The API server only receives configuration through the environment variables listed in its Deployment, and DriverPodLabels and DriverPodAnnotations were not among them, so anyone following the README would have set the keys and seen nothing happen.

This revision wires both keys into the API server Deployment, and into the PostgreSQL install as well, whose apiserver patch replaces the container environment list and therefore needs its own entries. I also added validation so that a label or annotation Kubernetes would reject now fails at startup with a clear message rather than surfacing later as a rejected driver pod. Annotation values stay free form, so the inline Istio proxy config keeps working.

I checked this on a real Kubernetes cluster rather than trusting the manifests alone. With the rendered Deployment applied and the ConfigMap holding the Istio values, the values came through into the API server container environment intact, escaped inline JSON and all. I also rendered every overlay to confirm the wiring is present in each, including multi user and PostgreSQL.

I am sorry that this costs the lgtm and another round of CI after you had both already signed off. @hbelmiro, whenever your schedule allows, I would be grateful for another look. @chensun, the manifests will still need your /approve when you have a moment. Thank you both for your patience.

juliusvonkohout commented on backend/README.md:
```diff
@@ -62,6 +62,35 @@ Run
+
+### Driver Pod Labels and Annotations
+
+The API server supports configuring custom labels and annotations for driver pods through the config file or the Kubernetes ConfigMap. This is useful for integration with service mesh (Istio), monitoring systems, or other infrastructure requirements.
```
```suggestion
The API server supports configuring custom labels and annotations for driver pods through the configuration file or the Kubernetes ConfigMap. This is useful for integration with service mesh (Istio), monitoring systems, or other infrastructure requirements.
```

juliusvonkohout commented on manifests/kustomize/base/installs/generic/postgres/pipeline-install-config.yaml:
We really need to get rid of this duplicate file in a follow up PR @kaikaila . manifests/kustomize/base/installs/generic/pipeline-install-config.yaml shoud be the single source of truth

juliusvonkohout commented on manifests/kustomize/base/postgresql/pipeline/ml-pipeline-apiserver-deployment-patch.yaml:
```diff
@@ -65,6 +65,20 @@ spec:
               optional: true
+        # Driver pod labels and annotations. This patch replaces the container env list,
+        # so these must be repeated here to reach the API server on PostgreSQL installs.
+        - name: DRIVERPODLABELS
```
```suggestion
        - name: DRIVER_POD_LABELS
```

juliusvonkohout commented on manifests/kustomize/base/postgresql/pipeline/ml-pipeline-apiserver-deployment-patch.yaml:
```diff
@@ -65,6 +65,20 @@ spec:
+          valueFrom:
+            configMapKeyRef:
+              name: pipeline-install-config
+              key: DriverPodLabels
```
```suggestion
              key: DRIVER_POD_LABELS
```

juliusvonkohout commented on manifests/kustomize/base/postgresql/pipeline/ml-pipeline-apiserver-deployment-patch.yaml:
```diff
@@ -65,6 +65,20 @@ spec:
+          valueFrom:
+            configMapKeyRef:
+              name: pipeline-install-config
+              key: DriverPodAnnotations
```
```suggestion
              key: DRIVER_POD_ANNOTATIONS
```

juliusvonkohout commented on manifests/kustomize/base/postgresql/pipeline/ml-pipeline-apiserver-deployment-patch.yaml:
```diff
@@ -65,6 +65,20 @@ spec:
+              name: pipeline-install-config
+              key: DriverPodLabels
+              optional: true
+        - name: DRIVERPODANNOTATIONS
```
```suggestion
        - name: DRIVER_POD_ANNOTATIONS
```

juliusvonkohout commented on manifests/kustomize/base/installs/generic/postgres/pipeline-install-config.yaml:
```diff
@@ -125,3 +125,9 @@ data:
+  ## Keys with the prefix "pipelines.kubeflow.org/" are reserved and will be filtered out.
+  ## Example: '{"sidecar.istio.io/inject":"true"}'
+  DriverPodLabels: ""
+  DriverPodAnnotations: ""
```
```suggestion
  DRIVER_POD_LABELS: ""
  DRIVER_POD_ANNOTATIONS: ""
```

thc1006 commented:
TLDR: every suggestion from @juliusvonkohout is applied, the keys are now DRIVER_POD_LABELS and DRIVER_POD_ANNOTATIONS everywhere, and two real defects that Copilot found are fixed, one of which would have silently disabled Istio injection.

@juliusvonkohout, thank you for taking the time to go through this, and for the concrete suggestions. I have applied all of them. The rename is now consistent across every place the keys appear, not only in the files you marked: the default install config, the PostgreSQL install config, the API server Deployment, the PostgreSQL apiserver patch, the Go configuration keys, and both README examples. The ConfigMap key, the environment variable, the Viper key and the config.json key are now one single name, which I believe is what you were after. Your wording fix in the README is in as well.

On the duplicate install config file, I have left it untouched and kept your note for the follow up PR you mentioned to @kaikaila, since it reaches well beyond this change.

Copilot raised three points and two of them were real, so I am glad it looked. The first one matters to operators. Viper does not reject a JSON null the way it rejects a boolean or a number. It quietly turns null into an empty string, so someone writing null for the Istio injection flag would have ended up with an empty label and no warning at all, believing injection was on. I reproduced that in a test before changing anything. Every value must now be a JSON string, and a null, a boolean, a number or a nested object all fail startup with a message naming the offending key. The second point was that GetDriverPodConfig took the read lock twice, which could hand back labels from one version of the cache and annotations from another, defeating the atomic snapshot that @hbelmiro asked for in his earlier review. Both maps are now copied under one lock.

Copilot's third point was that the compiler reaches into API server state. To be precise, Compile already read API server configuration before this PR, for the service account name, the CA bundle and the metadata TLS setting, so the coupling is not new. The suggestion is still the better shape, so the driver pod configuration now travels through argocompiler.Options in the same way as DefaultRunAsUser and DefaultHostUsers, and the standalone compiler receives nil explicitly rather than by accident.

@juliusvonkohout, whenever your schedule allows, I would be grateful if you could take another look. @hbelmiro, the lgtm was dropped by these commits, so it would need your /lgtm again when you have a moment. Thank you both.

thc1006 commented:
The two MLflow checks that are red are both artifactStorage=file variants, and the same ones are currently failing on master as well (four of them on 89d2530e8) with the identical SeaweedFS socket errors and read timeouts, so they do not appear to be caused by this change. I am of course happy to look again if you see it differently.

juliusvonkohout commented:
/retest

hbelmiro reviewed (approved):
/lgtm
/approve

juliusvonkohout commented:
/approve
